Generalized mime type support

a-ignatov-parc opened the pull request:
<!-- Enter details of the change here. Include additional tests that have been done, reference to the issue for tracking, etc. -->

Address https://github.com/OpenAPITools/openapi-generator/issues/17282

<!-- Please check the completed items below -->
### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh ./bin/configs/*.yaml
  ./bin/utils/export_docs_generators.sh
  ``` 
  (For Windows users, please run the script in [Git BASH](https://gitforwindows.org/))
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  IMPORTANT: Do **NOT** purge/delete any folders/files (e.g. tests) when regenerating the samples as manually written tests may be removed.
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master` (upcoming 7.1.0 minor release - breaking changes with fallbacks), `8.0.x` (breaking changes without fallbacks)
- [x] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.

---
@TiFu (2017/07) @taxpon (2017/07) @sebastianhaas (2017/07) @kenisteward (2017/07) @Vrolijkx (2017/09) @macjohnny (2018/01) @topce (2018/10) @akehir (2019/07) @petejohansonxo (2019/11) @amakhrov (2020/02) @davidgamero (2022/03) @mkusaka (2022/04)

macjohnny commented on modules/openapi-generator/src/main/resources/typescript/model/ObjectSerializer.mustache:
```diff
@@ -59,6 +49,45 @@ let typeMap: {[index: string]: any} = {
     {{/models}}
 }
 
+type MimeTypeDescriptor = {
```
can you please add docs to these types/properties/functions?

a-ignatov-parc replied:
@macjohnny, sure! Just to clarify, am I correct in understanding that we're talking about comments in the code, not some other docs, right?

macjohnny replied:
Yes, code comments are fine

a-ignatov-parc replied:
Done

macjohnny commented on modules/openapi-generator/src/main/resources/typescript/model/ObjectSerializer.mustache:
```diff
@@ -204,14 +233,15 @@ export class ObjectSerializer {
-            }
-        }
+
+        const supportedAndOrderedMediaTypes = supportedMimeTypePredicatesWithPriority.reduce<string[]>((result, predicate) => {
```
since only `supportedAndOrderedMediaTypes[0]` will be used, how about a for loop with early break on the first result instead of compiling the entire list to simplify the code a bit?

a-ignatov-parc replied:
@macjohnny, that's a fair point. I'll update the implementation. Thanks! 👍

a-ignatov-parc replied:
Updated

macjohnny commented on modules/openapi-generator/src/main/resources/typescript/model/ObjectSerializer.mustache:
```diff
@@ -59,6 +49,45 @@ let typeMap: {[index: string]: any} = {
+const isOctetStreamMimeType = mimeTypeSimplePredicateFactory('application', 'octet-stream');
+const isFormUrlencodedMimeType = mimeTypeSimplePredicateFactory('application', 'x-www-form-urlencoded');
+
+const supportedMimeTypePredicatesWithPriority: MimeTypePredicate[] = [
```
could you please add some tests to https://github.com/OpenAPITools/openapi-generator/tree/master/samples/openapi3/client/petstore/typescript/tests?

a-ignatov-parc replied:
@macjohnny, understood! Will do so 👍

a-ignatov-parc replied:
Done

macjohnny commented:
@a-ignatov-parc thanks for your contribution!

a-ignatov-parc commented:
@macjohnny, I apologize that addressing the feedback took longer than I initially anticipated. However, the PR is ready for the next round of review 👍

bodograumann commented:
Hey, I just wanted to say thanks for contributing this, @a-ignatov-parc ! Just when we needed it, all we had to do was update the generator :heart:

a-ignatov-parc commented:
@bodograumann, thank you for the feedback. It's great to know when your contribution has been useful to someone 🙂
